manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,431 @@
+        return int(self._web3().eth.coinbase)
+
+
+####################################################################################################
+
+
+# sam.moelius: If we decide to cache results returned from a RemoteWorldState, then they should NOT
+# be cached within an overlay.  The reason is that this could affect the results of subsequent
+# operations.  Consider a call to get_storage_data followed by a call to has_storage.  If nothing
+# was written to storage within the overlay, then the call to has_storage will throw an exception.
+# But if the result of the call to get_storage_data was cached in the overlay, then no exception
+# would be thrown.
+
+
+class OverlayWorldState(WorldState):
+    def __init__(self, underlay: WorldState):
+        self._underlay: WorldState = underlay
+        self._deleted_accounts: Set[int] = set()
+        self._nonce: Dict[int, Union[int, BitVec]] = {}
+        self._balance: Dict[int, Union[int, BitVec]] = {}
+        self._storage: Dict[int, Storage] = {}
+        self._code: Dict[int, Union[bytes, Array]] = {}
+        self._blocknumber: Optional[Union[int, BitVec]] = None
+        self._timestamp: Optional[Union[int, BitVec]] = None
+        self._difficulty: Optional[Union[int, BitVec]] = None
+        self._gaslimit: Optional[Union[int, BitVec]] = None
+        self._coinbase: Optional[Union[int, BitVec]] = None
+
+    def accounts(self) -> Set[int]:
+        accounts: Set[int] = set()
+        try:
+            accounts = self._underlay.accounts()
+        except NotImplementedError:
+            pass
+        return (
+            accounts
+            | self._nonce.keys()
+            | self._balance.keys()
+            | self._storage.keys()
+            | self._code.keys()
+        )
+
+    def get_nonce(self, address: int) -> Union[int, BitVec]:
+        if address in self._nonce:
+            return self._nonce[address]
+        else:
+            return self._underlay.get_nonce(address)
+
+    def get_balance(self, address: int) -> Union[int, BitVec]:
+        if address in self._balance:
+            return self._balance[address]
+        else:
+            return self._underlay.get_balance(address)
+
+    def has_storage(self, address: int) -> bool:
+        dirty = False
+        try:
+            dirty = self._underlay.has_storage(address)
+        except NotImplementedError:
+            pass
+        storage = self._storage.get(address)
+        if storage is not None:
+            dirty = dirty or storage.dirty
+        return dirty
+
+    def get_storage(self, address: int) -> Optional[Storage]:
+        storage = None
+        try:
+            storage = self._underlay.get_storage(address)
+        except NotImplementedError:
+            pass
+        # sam.moelius: Rightfully, the overlay's storage should be merged into the underlay's
+        # storage.  However, this is not currently implemented.
+        if storage is not None:
+            raise NotImplementedError
```

Comment:
The comment is helpful, but as it's on a different line, let's add a message to this error so that users will have some idea of what's missing if they encounter this error.

Reply:
I added such a message.